pkg/destinations/redshift/query.go

```diff
@@ -0,0 +1,108 @@
+package redshift
+
+import (
+	"database/sql"
+	"encoding/csv"
+	"encoding/json"
+	"fmt"
+	"io"
+)
+
+// QueryJSON executes a query on RedShift and writes results in JSON format to the provided writer.
+func (r *RedshiftServer) QueryJSON(query string, writer io.Writer) error {
+	rows, err := r.query(query)
+	if err != nil {
+		return err
+	}
+	defer rows.Close()
+
+	columns, err := rows.Columns()
+	if err != nil {
+		return err
+	}
+
+	result := make([]map[string]interface{}, 0)
+	values := make([]interface{}, len(columns))
+	valuePtrs := make([]interface{}, len(columns))
+
+	for rows.Next() {
+		for i := range columns {
+			valuePtrs[i] = &values[i]
+		}
+		if err := rows.Scan(valuePtrs...); err != nil {
+			return err
+		}
+
+		row := make(map[string]interface{})
+		for i, col := range columns {
+			var v interface{}
+			val := values[i]
+			b, ok := val.([]byte)
+			if ok {
+				v = string(b)
+			} else {
+				v = val
+			}
+			row[col] = v
+		}
+		result = append(result, row)
```

Comment:
This is appending the entire result set to a slice in memory. We do not want to do that since the result set could be very large. Instead, we want to stream data as the results are being generated.

From the issue #117:

> Data should be streamed to the writer and avoid allocations as much as possible. Do not buffer all data into memory before returning to the client. Do not generate the entire JSON/CSV payload in memory.